lib/std/crypto/sha2.zig

```diff
@@ -1,87 +1,114 @@
+//! Secure Hashing Algorithm 2 (SHA2)
+//!
+//! A suite of 5 secure hash functions with varying digest lengths:
+//! * Sha224
+//! * Sha256
+//! * Sha384
+//! * Sha512
+//! * Sha512/digest_len, where 1 < `digest_len` < 512 and `digest_len` != 384
+//!
+//! Published by the National Institue of Standards and Technology (NIST):
+//! https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf
+//! https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-208.pdf
+
 const std = @import("../std.zig");
 const builtin = @import("builtin");
 const mem = std.mem;
 const math = std.math;
 const htest = @import("test.zig");
 
-/////////////////////
-// Sha224 + Sha256
-
-const RoundParam256 = struct {
-    a: usize,
-    b: usize,
-    c: usize,
-    d: usize,
-    e: usize,
-    f: usize,
-    g: usize,
-    h: usize,
-    i: usize,
+pub const Sha224 = Sha2x32(iv224, 224);
+pub const Sha256 = Sha2x32(iv256, 256);
+pub const Sha384 = Sha2x64(iv384, 384);
+pub const Sha512 = Sha2x64(iv512, 512);
+
+// Truncated variants that save space but reduce security.
+// T = original IV
+// _ = changed IV
+/// SHA-256 truncated to 192 bits per FIPS 800-208.
+pub const Sha256T192 = Sha2x32(iv256, 256);
+
+/// SHA-512 truncated to 224 bits per FIPS 180.
+pub const Sha512_224 = Sha512Truncated(224, IvStrategy.change);
+/// SHA-512 truncated to 224 bits.
+pub const Sha512T224 = Sha512Truncated(224, IvStrategy.keep);
+
+/// SHA-512 truncated to 256 bits per FIPS 180.
+pub const Sha512_256 = Sha512Truncated(256, IvStrategy.change);
+/// SHA-512 truncated to 256 bits.
+pub const Sha512T256 = Sha512Truncated(256, IvStrategy.keep);
+
+const IvStrategy = enum {
+    /// Change the IV per NIST FIPS 180 (released 2012).
+    change,
+    /// Use the original IV (released 2001).
+    keep,
 };
-
-fn roundParam256(a: usize, b: usize, c: usize, d: usize, e: usize, f: usize, g: usize, h: usize, i: usize) RoundParam256 {
-    return RoundParam256{
-        .a = a,
-        .b = b,
-        .c = c,
-        .d = d,
-        .e = e,
-        .f = f,
-        .g = g,
-        .h = h,
-        .i = i,
+/// SHA-512 truncated to length `digest_bits` using `iv_strategy`.
+/// This may be more performant than SHA224 and SHA256.
+fn Sha512Truncated(digest_bits: comptime_int, comptime iv_strategy: IvStrategy) type {
```

Comment:
I'm still confused about why there is `Sha512Truncated` but not `Sha256Truncated` (for SHA-224 and SHA-192).

Reply:
`Sha2x32` and `Sha2x64` already have `digest_bits` params. I've renamed this to `Sha512Advanced`.